### PR TITLE
Add CI for s390x and fix big-endian arch

### LINKED ISSUE #97: Testsuite fails on s390x with GHC 9.0.2
Testsuite fails on s390x, ppc64 and sparc64 with GHC 9.0.2 (see https://buildd.debian.org/status/logs.php?pkg=haskell-unicode-data&ver=0.3.1-1) with the following error:
```
Unicode.Char
  Unicode blocks
    Characters not in any block are unassigned
    Characters are in the definition of their corresponding block FAILED [1]
    Characters in a block definition have the corresponding block FAILED [2]
  Unicode general categories
    generalCategory
      # PENDING: Incompatible GHC Unicode version
  Character classification
    isAlpha
      # PENDING: Incompatible GHC Unicode version
    isAlphaNum
      # PENDING: Incompatible GHC Unicode version
    isControl
      # PENDING: Incompatible GHC Unicode version
    isLetter
      # PENDING: Incompatible GHC Unicode version
    isMark
      # PENDING: Incompatible GHC Unicode version
    isPrint
      # PENDING: Incompatible GHC Unicode version
    isPunctuation
      # PENDING: Incompatible GHC Unicode version
    isSeparator
      # PENDING: Incompatible GHC Unicode version
    isSpace
      # PENDING: Incompatible GHC Unicode version
    isSymbol
      # PENDING: Incompatible GHC Unicode version
  Case
    isLower
      # PENDING: Incompatible GHC Unicode version
    isUpper
      # PENDING: Incompatible GHC Unicode version
    toLower
      # PENDING: Incompatible GHC Unicode version
    toLowerString
      Examples
      Common mapping should match simple one
        # PENDING: Incompatible GHC Unicode version
      Idempotency of 'foldMap toLowerString'
    toUpper
      # PENDING: Incompatible GHC Unicode version
    toUpperString
      Examples
      Common mapping should match simple one
        # PENDING: Incompatible GHC Unicode version
      Idempotency of 'foldMap toUpperString'
    toTitle
      # PENDING: Incompatible GHC Unicode version
    toTitleString
      Examples
      Common mapping should match simple one
        # PENDING: Incompatible GHC Unicode version
    toCaseFoldString
      Examples
      Idempotency of 'foldMap toCaseFoldString'
  Numeric
    isNumber
      # PENDING: Incompatible GHC Unicode version
    isNumber implies a numeric value

Failures:

  test/Unicode/CharSpec.hs:65:30: 
  1) Unicode.Char, Unicode blocks, Characters are in the definition of their corresponding block
       Character “'\128'” is not in the block “BasicLatin”.

  To rerun use: --match "/Unicode.Char/Unicode blocks/Characters are in the definition of their corresponding block/"

  test/Unicode/CharSpec.hs:75:22: 
  2) Unicode.Char, Unicode blocks, Characters in a block definition have the corresponding block
       Block is different for “'\128'”. Expected: “Just Latin1Supplement” but got: “Just BasicLatin”.

  To rerun use: --match "/Unicode.Char/Unicode blocks/Characters in a block definition have the corresponding block/"
```

The previous version of this package (0.3.0) built fine on these architectures (see https://buildd.debian.org/status/logs.php?pkg=haskell-unicode-data&ver=0.3.0-2). It appears version 0.3.1 is broken on 64-bit big-endian architectures. Can you please help us identify the problem?

### PR DESCRIPTION
**Pull request by wismill**
Fix for 32bits words on big-endian arch.

Fixes #97

### CODE REVIEW

**Review by harendra-kumar — CHANGES_REQUESTED** [review]
We need to fix the following files as well for the same issue:

* unicode-data-scripts/test/Unicode/Char/General/ScriptsSpec.hs
* unicode-data-scripts/lib/Unicode/Char/General/Scripts.hs

We may have problems in future if we forget to include MachDeps.h , this is a nasty thing that is not caught by any tool and can be easily missed. We can do two things to mitigate this:

* define a CPP macro for reading Word32 with endianness handling and then use that macro everywhere. That way we need to include MachDeps only in the file defining that macro. Both byteswap and narrowing can be included in that macro so we do not miss it anywhere.
* Add a CI for s390x. We can copy it from unicode-transforms repo.

**Comment by wismill** [comment]
> We need to fix the following files as well for the same issue:
> 
>     * unicode-data-scripts/test/Unicode/Char/General/ScriptsSpec.hs
> 
>     * unicode-data-scripts/lib/Unicode/Char/General/Scripts.hs
> 

These are already fixed.

> We may have problems in future if we forget to include MachDeps.h , this is a nasty thing that is not caught by any tool and can be easily missed. We can do two things to mitigate this:
> 
>     * define a CPP macro for reading Word32 with endianness handling and then use that macro everywhere. That way we need to include MachDeps only in the file defining that macro. Both byteswap and narrowing can be included in that macro so we do not miss it anywhere.
> 
>     * Add a CI for s390x. We can copy it from unicode-transforms repo.

I think the latter is better. I am giving a try.

File: unicode-data-names/lib/Unicode/Internal/Char/UnicodeData/DerivedName.hs [hunk 1]
```diff
@@ -51,9 +51,9 @@ name (C# c#) = getName 0# 149185#
     indexInt32OffAddr'# k# =
 #ifdef WORDS_BIGENDIAN
 #if MIN_VERSION_base(4,16,0)
-        word2Int# (byteSwap32# (word32ToWord# (indexWord32OffAddr# offsets# k#)))
+        word2Int# (narrow32Word# (byteSwap32# (word32ToWord# (indexWord32OffAddr# offsets# k#))))
```
[the remark this answers is not in the document]
Comment (wismill) [reply unicode-data-names/lib/Unicode/Internal/Char/UnicodeData/DerivedName.hs:54]: I did not find much information. I found some hints in `cryptonite` package.

My understanding is that _byte swapping_ adapt the endianness and is not limited to 2 bytes. Here `byteSwap32#` is for 4 bytes, and `byteSwap16#` would be for 2, etc.

**Comment by wismill** [comment]
@harendra-kumar there is an issue with the CI for GHC 8.0.2 and 8.2.2. It seems they are not supported anymore.

**Comment by harendra-kumar** [comment]
> > We need to fix the following files as well for the same issue:
> > ```
> > * unicode-data-scripts/test/Unicode/Char/General/ScriptsSpec.hs
> > 
> > * unicode-data-scripts/lib/Unicode/Char/General/Scripts.hs
> > ```
> 
> These are already fixed.

It see this in `unicode-data-scripts/lib/Unicode/Char/General/Scripts.hs`:

```
#ifdef WORDS_BIGENDIAN
#if MIN_VERSION_base(4,16,0)
            byteSwap32# (word32ToWord# (indexWord32OffAddr# addr# k#));
#else
            byteSwap32# (indexWord32OffAddr# addr# k#);
#endif
```

Don't we need a call to narrowWord here?

**Comment by harendra-kumar** [comment]
> @harendra-kumar there is an issue with the CI for GHC 8.0.2 and 8.2.2. It seems they are not supported anymore.

The installation seems to be failing, not sure what's wrong. We can investigate if someone asks for it, otherwise let's ignore it for now.

File: .github/workflows/s390x.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  # and even so `ghc -O` is prohibitively expensive.
+  build:
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - uses: uraimo/run-on-arch-action@v2.1.1
+      timeout-minutes: 60
+      with:
+        arch: s390x
+        distro: ubuntu20.04
```
[the remark this answers is not in the document]
Comment (wismill) [reply .github/workflows/s390x.yml:27]: Yes, that's what I did in #99

**Comment by wismill** [comment]
After the experiment in #99, I think this is ready to merge if the relevant tests pass (let’s discard GHC ≤ 8.2). Keep in mind that we can fully test only against GHC 9.6. And GHC support for Unicode < 12.1 is not reliable, because there are some discrepancies (I checked manually some errors).

**Comment by harendra-kumar** [comment]
It seems you missed this https://github.com/composewell/unicode-data/pull/98#issuecomment-1336451077 . Can you please respond to this.

**Comment by wismill** [comment]
> It seems you missed this [#98 (comment)](https://github.com/composewell/unicode-data/pull/98#issuecomment-1336451077) . Can you please respond to this.

I think I did. Are you using the latest version?

**Comment by wismill** [comment]
@harendra-kumar kind reminder

**Review by harendra-kumar — APPROVED** [review]

**Comment by harendra-kumar** [comment]
> I think I did fix it. Are you using the latest version?

Looks like I missed it, maybe I did not pull it again or ignored the pull error.

**Pull request merged**